### PR TITLE
Validate repository configuration changes on PR creation

### PR DESCRIPTION
**Pull request by marcos-arranz**
[Linear ticket](https://linear.app/pleo/issue/DX-593/ensure-app-validates-repository-changes-that-modify-repository)

Validates an incoming PR event, and validates if within the files changed there is a template config.

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -15,6 +23,60 @@ const findBranchesToProcess = () => {
...
+    repo: payload.repository.name,
+  }
+  app.log(`Pull request event happened on '${context.payload.pull_request}'`)
+
+  try {
+    const filesChanged = await context.octokit.pulls.listFiles({
+      ...repository,
+      pull_number: payload.number
+    })
+    const configFile = filesChanged.data.find(x => x.filename === configFileName)
```
Comment (andersfischernielsen) [inline src/app.ts:39]: I'd give this a more telling name 😄 

```suggestion
    const configFile = filesChanged.data.find(file => file.filename === configFileName)
```

File: src/app.ts [hunk 2]
```diff
@@ -15,6 +23,60 @@ const findBranchesToProcess = () => {
...
+    if (configFile) {
+
```
Comment (andersfischernielsen) [inline src/app.ts:41]: If you run `yarn format` these formatting issues go away 😄

File: src/app.ts [hunk 3]
```diff
@@ -15,6 +23,60 @@ const findBranchesToProcess = () => {
   return new RegExp(branches)
 }
 
+const processPullRequest = () => async (context: Context<'pull_request'>, app: Probot) => {
```
Comment (andersfischernielsen) [inline src/app.ts:26]: We can avoid passing in the entire `Probot` object in function calls by just taking the Pino `Logger`. That'll also follow the logic in the rest of the app implementation for consistency 😄 

```suggestion
const processPullRequest = (log: Logger) => async (context: Context<'pull_request'>) => {
```

File: src/types.ts [hunk 4]
```diff
@@ -35,5 +35,32 @@ export interface PRDetails {
...
+  source: string,
+  destination: string
+}
+
+export interface TemplateConfig {
+  version: string,
+  automerge: boolean,
+  files: Array<File>,
+  values: object
+}
```
Comment (andersfischernielsen) [inline src/types.ts:48]: Isn't this the same as `RepositoryConfiguration`? 🤔

File: src/types.ts [hunk 5]
```diff
@@ -35,5 +35,32 @@ export interface PRDetails {
   title: string
   description: string
 }
+interface File {
+  source: string,
+  destination: string
+}
```
Comment (andersfischernielsen) [inline src/types.ts:41]: Isn't this the same as `PathConfiguration`?

File: src/app.ts [hunk 6]
```diff
@@ -1,9 +1,17 @@
...
 import { config } from 'dotenv'
 import { determineConfigurationChanges } from './configuration'
 import { renderTemplates } from './templates'
 import { commitFiles, getCommitFiles } from './git'
+import validator from "./schema-validator";
+import schema from './template-schema.json'
+import { JSONSchemaType } from "ajv";
+import { TemplateConfig } from './types'
+import { createCheckRun, resolveCheckRun } from './github'
+
```
Comment (andersfischernielsen) [inline src/app.ts:12]: `yarn format` will take care of this too 😄

File: src/app.ts [hunk 7]
```diff
@@ -15,6 +23,60 @@ const findBranchesToProcess = () => {
...
+      app.log.debug(`Found file ${configFile.filename}`)   
+      const checkRun = await createCheckRun(
+        context.octokit,
+        {
+          owner: repository.owner,
+          repo: repository.repo,
+          sha: payload.pull_request.head.sha
+        }
+      )
```
Comment (andersfischernielsen) [inline src/app.ts:50]: I'd extract the object out here for readability: 

```suggestion
      const toCreate = {
          owner: repository.owner,
          repo: repository.repo,
          sha: payload.pull_request.head.sha
        }
      const createdCheck = await createCheckRun(context.octokit, toCreate)
```

File: src/app.ts [hunk 8]
```diff
@@ -15,6 +23,60 @@ const findBranchesToProcess = () => {
...
+        path: configFile.filename,
+        ref: payload.pull_request.head.ref
+      })
+
+      const { content } = fileContent.data as { content: string }
+      const decodedContent = Buffer.from(content, 'base64').toString()
+      app.log.debug(`decoded content ${decodedContent}`)
+
+
+      const result = validator(schema as JSONSchemaType<TemplateConfig>, decodedContent)
```
Comment (andersfischernielsen) [inline src/app.ts:63]: I'd rip the cast out here to make it explicit that we're casting

```suggestion
      const typedSchema = schema as JSONSchemaType<TemplateConfig>
      const result = validator(typedSchema, decodedContent)
```

File: src/app.ts [hunk 9]
```diff
@@ -15,6 +23,60 @@ const findBranchesToProcess = () => {
...
+      const resultString = (result) ? "success" : "failure"
+
+      await resolveCheckRun(context.octokit, {
+        owner: repository.owner,
+        repo: repository.repo,
+        sha: context.payload.pull_request.head.sha,
+        result: resultString, 
+        check_run_id: checkRun.data.id
+      })
```
Comment (andersfischernielsen) [inline src/app.ts:72]: I'd also rip this out to make the invocation cleaner: 

```suggestion
      const resolveCheck = {
        owner: repository.owner,
        repo: repository.repo,
        sha: context.payload.pull_request.head.sha,
        result: resultString, 
        check_run_id: checkRun.data.id
      }
      await resolveCheckRun(context.octokit, resolveCheck)
```

File: src/schema-validator.ts [hunk 10]
```diff
@@ -0,0 +1,11 @@
+import Ajv, {JSONSchemaType} from "ajv";
+import { parse } from "yaml";
+import { TemplateConfig } from "./types";
+
+export default (schema: JSONSchemaType<TemplateConfig>, input: string) => {
+    const ajv = new Ajv() 
+    const validate = ajv.compile(schema)
+    const valid = validate(parse(input))
+    if (!valid) console.log(validate.errors)
```
Comment (andersfischernielsen) [inline src/schema-validator.ts:9]: We should probably return the validation errors to give to teams so they know what's wrong with their configuration 😄 WDYT? 

We could return a 
```
{
  valid: boolean,
  errors?: string[]
}

type that's easy to present in a PR comment or something similar 😄

File: src/types.ts [hunk 11]
```diff
@@ -35,5 +35,32 @@ export interface PRDetails {
...
+  sha: string
+}
+
+export interface UpdateCheckInput {
+  owner: string,
+  repo: string,
+  sha: string,
+  result: string,
+  check_run_id: number
+}
```
Comment (andersfischernielsen) [inline src/types.ts:62]: We could merge these into a single `Check` type:

```suggestion
export interface Check {
  owner: string,
  repo: string,
  sha: string,
  result?: string,
  check_run_id?: number
}
```

Comment (marcos-arranz) [reply src/types.ts:62]: I'd rather extend it, the `result` and `check_run_id` are not optional on the UpdateCheckInput

File: src/app.ts [hunk 12]
```diff
@@ -15,6 +23,60 @@ const findBranchesToProcess = () => {
...
+
+      const { content } = fileContent.data as { content: string }
+      const decodedContent = Buffer.from(content, 'base64').toString()
+      app.log.debug(`decoded content ${decodedContent}`)
+
+
+      const result = validator(schema as JSONSchemaType<TemplateConfig>, decodedContent)
+      const resultString = (result) ? "success" : "failure"
+
+      await resolveCheckRun(context.octokit, {
```
[the remark this answers is not in the document]
Comment (marcos-arranz) [reply src/app.ts:66]: yeah, all the error handling is missing at the moment

File: src/schema-validator.ts [hunk 13]
```diff
@@ -0,0 +1,11 @@
+import Ajv, {JSONSchemaType} from "ajv";
+import { parse } from "yaml";
+import { TemplateConfig } from "./types";
+
+export default (schema: JSONSchemaType<TemplateConfig>, input: string) => {
```
[the remark this answers is not in the document]
Comment (marcos-arranz) [reply src/schema-validator.ts:5]: makes sense, I didn't know. Thanks!

File: src/types.ts (hunk 4 above)
Comment (marcos-arranz) [reply src/types.ts:48]: yes, I missed it. Changing it.

File: src/types.ts (hunk 5 above)
Comment (marcos-arranz) [reply src/types.ts:41]: Same here, changing it

File: src/app.ts (hunk 2 above)
Comment (andersfischernielsen) [reply src/app.ts:41]: Auto-format on save should be enabled in the VSCode settings of the repository too - maybe they're disabled

File: src/types.ts (hunk 11 above)
Comment (andersfischernielsen) [reply src/types.ts:62]: Ah, you could do this then: 

```typescript
export interface CheckInput {
  owner: string,
  repo: string,
  sha: string,
}

export interface CheckOutput extends CheckInput {
  result?: string,
  check_run_id?: number
}
```

or via types (whichever you prefer):

```typescript
export type CheckInput {
  owner: string,
  repo: string,
  sha: string,
}

export type CheckOutput & CheckInput {
  result?: string,
  check_run_id?: number
}
```

File: src/app.ts (hunk 3 above)
Comment (andersfischernielsen) [reply src/app.ts:26]: Ah, yeah - that'd be optimal 😄

**Comment by marcos-arranz** [comment]
Thanks @andersfischernielsen, let me go through every comment and I will send a new version including the error handling

File: src/app.ts (hunk 2 above)
Comment (marcos-arranz) [reply src/app.ts:41]: Just added it to my VS code settings. It wasn't enabled

File: src/app.ts [hunk 14]
```diff
@@ -1,9 +1,19 @@
...
-import { commitFiles, getCommitFiles } from './git'
+import {
+  approvePullRequestChanges,
+  commitFiles,
+  getCommitFiles,
+  getFilesChanged,
+  requestPullRequestChanges,
+} from './git'
+import { validateTemplateConfiguration } from './schema-validator'
+import { createCheckRun, resolveCheckRun } from './checks'
```
[the remark this answers is not in the document]
Comment (marcos-arranz) [reply src/app.ts:14]: makes sense, but I will change it in a different PR

File: src/app.ts [hunk 15]
```diff
@@ -21,6 +31,67 @@ const extractRepositoryInformation = (payload: PushEvent) => {
...
+      ...repository,
+      path: configFile,
+      ref,
+    })
+
+    const { content } = fileContent.data as { content: string }
+    const decodedContent = Buffer.from(content, 'base64').toString()
+    log.debug(`Saw configuration file contents:`)
+    log.debug(decodedContent)
+    const { result, errors } = validateTemplateConfiguration(decodedContent)(log)
```
[the remark this answers is not in the document]
Comment (marcos-arranz) [reply src/app.ts:75]: Will deal properly with the messaging and check resolution on a different PR

**Review by andersfischernielsen — APPROVED** [review]
LGTM! Nice!

**Pull request merged**